**EQUIPO 51**

Elliot Nez Arredondo | A01796107

Luis Antonio Calderon Mata | A00278401


Fernando Guzman Briones | A01039274

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [12]:
# Incluye todas las librerías que consideres adecuadas:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas import read_csv
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_curve, auc, mean_squared_error
import matplotlib.pyplot as plt
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.dummy import DummyClassifier
from sklearn.metrics import fbeta_score, make_scorer
from imblearn.metrics import geometric_mean_score
import statistics # import the statistics module from the standard library
from numpy import std
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_validate
from statistics import mean
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc, precision_recall_curve
from sklearn.metrics import roc_auc_score, average_precision_score
import time # Import the time module

# Define MAPE function once globally
def mean_absolute_percentage_error(y_true, y_pred):
    epsilon = np.finfo(np.float64).eps
    return np.mean(np.abs((y_true - y_pred) / np.maximum(y_true, epsilon))) * 100

# Initialize a list to store metrics for all models globally
all_model_metrics = []


In [13]:
# Re-executing Random Forest Regressor to collect metrics
import time
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Filtrar filas de resumen y valores nulos
df_model = df.dropna(subset=['PROJECT NAME']).copy()

# 2. Preprocesamiento
features = ['District', 'Region', 'EQUIP TYPE', 'PROD CODE', 'Month', 'Plan -2 Costs']
X = pd.get_dummies(df_model[features], drop_first=True)
y = df_model['MV']

# 3. Modelo
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
rf = RandomForestRegressor(n_estimators=100, random_state=42)

start_time = time.time()
rf.fit(X_train, y_train)
y_pred = rf.predict(X_test)
end_time = time.time()
rf_execution_time = end_time - start_time

# 4. Evaluación
rf_r2 = r2_score(y_test, y_pred)
rf_mae = mean_absolute_error(y_test, y_pred)
rf_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
rf_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"R2 Score: {rf_r2:.4f}")
print(f"MAE: {rf_mae:.2f}")
print(f"RMSE: {rf_rmse:.2f}")
print(f"MAPE: {rf_mape:.2f}%")
print(f"Execution Time: {rf_execution_time:.2f} seconds")

# Store metrics
all_model_metrics.append({
    'Model': 'Random Forest Regressor',
    'R2 Score': rf_r2,
    'MAE': rf_mae,
    'RMSE': rf_rmse,
    'MAPE (%)': rf_mape,
    'Execution Time (s)': rf_execution_time
})


KeyError: "['Plan -2 Costs'] not in index"

In [6]:
path = '/content/drive/MyDrive/Proyecto Integrador/NBS_Mod_QR Orders December 2025(OERP (NBS)).csv'
df = read_csv(path, encoding='latin1') # Removed header=None to let pandas infer header

FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Proyecto Integrador/NBS_Mod_QR Orders December 2025(OERP (NBS)).csv'

In [ ]:
#luis
import os
DIR = "/content/drive/MyDrive/Colab Notebooks/MNA/Proyecto Integrador"
os.chdir(DIR)
df = pd.read_csv('NBS_Mod_QR Orders December 2025.csv')

In [7]:
from google.colab import drive
drive.mount('/content/drive')

import os
import pandas as pd

DIR = "/content/drive/MyDrive/Colab Notebooks/PIntegrador2026"
os.chdir(DIR)
df = pd.read_csv('NBS_Mod_QR Orders December 2025')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Colab Notebooks/PIntegrador2026'

In [10]:
from google.colab import drive
drive.mount('/content/drive')

import os
import pandas as pd

DIR = "/content/drive/MyDrive/Colab Notebooks/PIntegrador26"
os.chdir(DIR)

print(os.listdir())
df = pd.read_csv('NBS_Mod_QR Orders December 2025.csv')

df.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
['NBS_Mod_QR Orders December 2025.csv']


,Group,District,Region,Date,Month,Region Vlookup,District Vlookup,Branch Vlookup,Change Order Lookup,PROJECT,...,Unnamed: 35,Plan -2 Costs,Plan -2 CMII $,Plan -2 CMII%,MP?,CMII %,Change Order?,Product,DISTRICT,Branch Name
0,10100.0,Central West,West,12/31/2025,12.0,ALL REGIONSAll OrdersAD12,Central WestAll OrdersAD12,10100All OrdersAD12,All Orders,6741283.0,...,NaN,NaN,NaN,NaN,No,40.74%,Yes,Addwork,DMD,San Antonio
1,10414.0,Central,East,12/31/2025,12.0,ALL REGIONSAll OrdersNE12,CentralAll OrdersNE12,10414All OrdersNE12,All Orders,6741284.0,...,NaN,NaN,NaN,NaN,No,100.00%,Yes,MonoSpace 500 2-3 landing(ENA),DCE,Nashville
2,10414.0,Central,East,12/31/2025,12.0,ALL REGIONSAll OrdersNE12,CentralAll OrdersNE12,10414All OrdersNE12,All Orders,6741284.0,...,NaN,NaN,NaN,NaN,No,100.00%,Yes,MonoSpace 500 2-3 landing(ENA),DCE,Nashville
3,10100.0,Central West,West,12/31/2025,12.0,ALL REGIONSAll OrdersAD12,Central WestAll OrdersAD12,10100All OrdersAD12,All Orders,6753566.0,...,NaN,NaN,NaN,NaN,No,46.05%,Yes,Addwork,DMD,San Antonio
4,10260.0,Central,East,12/31/2025,12.0,ALL REGIONSAll OrdersTM12,CentralAll OrdersTM12,10260All OrdersTM12,All Orders,6772984.0,...,NaN,NaN,NaN,NaN,No,68.71%,Yes,Time and material,DCE,Grand Rapids


In [ ]:
df.shape[0]

14582

#**Modelo Alternativo 1 - Random Forest Regressor**

In [14]:
if 'Change Order?' in df_clean.columns:
    change_order_counts = df_clean['Change Order?'].value_counts()
    print("Counts for 'Change Order?':")
    print(change_order_counts)
else:
    print("Column 'Change Order?' not found in df_clean.")


NameError: name 'df_clean' is not defined

In [15]:
# Primero procedemos con una limpieza de datos

# Clean column names by stripping whitespace
df.columns = df.columns.str.strip()

# Make column names unique if there are duplicates after stripping
cols = pd.Series(df.columns)
duplicate_mask = cols.duplicated(keep=False)
if duplicate_mask.any():
    new_columns = []
    counts = {}
    for col_name in cols:
        if col_name in counts:
            counts[col_name] += 1
            new_columns.append(f'{col_name}.{counts[col_name]}')
        else:
            counts[col_name] = 0
            new_columns.append(col_name)
    df.columns = new_columns

# Correct specific column name 'CMII %' to 'CMII%' if it exists (this should be handled by stripping and uniqueness now, but kept for robustness)
if 'CMII %' in df.columns:
    df.rename(columns={'CMII %': 'CMII%'}, inplace=True)

# The DataFrame `df` is now expected to have correct headers already
# from the previous `read_csv` call in cell Cp4jBSMqNVNl.
# The following lines for manual header extraction are no longer needed.
# new_columns_series = df.iloc[0]
# df.columns = new_columns_series.astype(str).str.strip().tolist()
# df = df[1:].copy()

# objetivos a limpiar
# Update target_cols to reflect potential unique names if CMII% was duplicated.
# For now, let's keep the original names, assuming the uniqueness step will handle the DataFrame side.
target_cols = ['MV', 'CMII $', 'PLAN COSTS', 'CMII%']

def clean_currency_and_percent(val):
    # If val is an array-like object (e.g., a Series), it means the data is malformed.
    # Return NaN as it cannot be cleaned as a single currency/percentage string.
    if isinstance(val, (list, tuple, np.ndarray, pd.Series)):
        return np.nan

    # Handle NaN or empty string values first
    if pd.isna(val) or (isinstance(val, str) and val.strip() == '') or val == '-':
        # Treat '-' as 0.0 for accounting context, otherwise NaN for truly missing
        return 0.0 if val == '-' else np.nan

    if isinstance(val, str):
        val = val.strip()
        # Handle parentheses for negatives: (1,234.56) -> -1234.56
        if val.startswith('(') and val.endswith(')'):
            val = '-' + val[1:-1]
        # Remove symbols
        val = val.replace('$', '').replace('%', '').replace(',', '')
        try:
            return float(val)
        except ValueError:
            return np.nan
    else:
        # If it's not a string and not NaN/empty/array-like, it might be a number already.
        # Attempt to convert to float in case it's an int or other numeric type.
        try:
            return float(val)
        except (ValueError, TypeError):
            return np.nan # Unparseable non-string, non-NaN scalar

# Aplicar limpieza
for col in target_cols:
    if col in df.columns: # Check if column exists after potential renaming of duplicates
        df[col] = df[col].apply(clean_currency_and_percent)
    else:
        print(f"Warning: Column '{col}' not found after cleaning column names.")

#includio LUIS
# Eliminar registros donde MV sea NaN o vacío
# Eliminar símbolos comunes
df["MV"] = (
    df["MV"]
    .astype(str)
    .str.replace("$", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.strip()
)

# Convertir a numérico
df["MV"] = pd.to_numeric(df["MV"], errors="coerce")

# Eliminar filas donde 'Change Order' es 'Yes'
if 'Change Order?' in df.columns:
    df = df[df['Change Order?'] != 'Yes'].copy()
    print(f"Filtered out rows where 'Change Order?' is 'Yes'. New shape: {df.shape}")
else:
    print("Warning: 'Change Order?' column not found. Skipping filtering by 'Change Order?'.")

# Ver cuántos NaN hay ahora
print("NaN en MV:", df["MV"].isna().sum())

# Eliminar filas con NaN en MV
df = df.dropna(subset=["MV"])

# Resumen
# Filter target_cols to only include columns that actually exist in the DataFrame
existing_target_cols = [col for col in target_cols if col in df.columns]

stats = df[existing_target_cols].describe()
missing_values = df[existing_target_cols].isna().sum()

print("Descriptive Statistics after cleaning:")
print(stats)
print("\nMissing values per column:")
print(missing_values)

# datos limpios en un nuevo csv
cleaned_file_name = 'Cleaned_NBS_Orders_Targets.csv'
df.to_csv(cleaned_file_name, index=False)

print(f"\nCleaned file saved as: {cleaned_file_name}")
print("Shape final:", df.shape)


/tmp/ipykernel_1303/1163967771.py:70: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  df[col] = df[col].apply(clean_currency_and_percent)


Filtered out rows where 'Change Order?' is 'Yes'. New shape: (14582, 45)
NaN en MV: 4964
Descriptive Statistics after cleaning:
                 MV        CMII $    PLAN COSTS  CMII%  CMII%
count  9.618000e+03  7.804000e+03  9.473000e+03    0.0    0.0
mean   1.399279e+05  4.074681e+04  1.246566e+05    NaN    NaN
std    2.713520e+06  1.383335e+06  4.174359e+06    NaN    NaN
min    1.000000e-02  1.000000e-02  1.000000e-02    NaN    NaN
25%    4.283788e+03  1.387568e+03  2.813130e+03    NaN    NaN
50%    1.617610e+04  1.258351e+04  1.262408e+04    NaN    NaN
75%    1.553688e+05  2.838961e+04  1.320388e+05    NaN    NaN
max    2.295369e+08  1.205262e+08  4.061443e+08    NaN    NaN

Missing values per column:
MV               0
CMII $        1814
PLAN COSTS     145
CMII%         9618
CMII%         9618
dtype: int64

Cleaned file saved as: Cleaned_NBS_Orders_Targets.csv
Shape final: (9618, 45)


In [16]:
if 'Change Order?' in df.columns:
    print(f"Unique values for 'Change Order?': {df['Change Order?'].unique()}")
    print(f"Counts for 'Change Order?':\n{df['Change Order?'].value_counts()}")
    print(f"Data type of 'Change Order?': {df['Change Order?'].dtype}")
else:
    print("Column 'Change Order?' not found in the DataFrame.")

Unique values for 'Change Order?': [' Yes ' ' No ' nan]
Counts for 'Change Order?':
Change Order?
No      6083
Yes     3523
Name: count, dtype: int64
Data type of 'Change Order?': object


In [17]:
if 'Product' in df.columns:
    # Strip whitespace from 'Product' column values
    df['Product'] = df['Product'].str.strip()

    # List of product types to exclude
    excluded_products = ['Addwork', 'Project Expense', 'Others', 'Temp Use', 'Time and material']

    # Filter out rows where 'Product' column is in the excluded_products list
    initial_rows = df.shape[0]
    df = df[~df['Product'].isin(excluded_products)].copy()
    final_rows = df.shape[0]
    print(f"Removed {initial_rows - final_rows} rows based on 'Product' exclusion.")
    print(f"New DataFrame shape: {df.shape}")
else:
    print("Column 'Product' not found in the DataFrame. No filtering applied.")

Removed 4372 rows based on 'Product' exclusion.
New DataFrame shape: (5246, 45)


In [18]:
if 'Change Order?' in df.columns:
    # Strip whitespace from 'Change Order?' values before comparison
    df['Change Order?'] = df['Change Order?'].str.strip()
    df = df[df['Change Order?'] != 'Yes'].copy()
    print(f"Filtered out rows where 'Change Order?' is 'Yes'. New shape: {df.shape}")
    # Save the modified DataFrame to CSV
    cleaned_file_name = 'Cleaned_NBS_Orders_Targets.csv'
    df.to_csv(cleaned_file_name, index=False)
    print(f"Updated DataFrame saved as: {cleaned_file_name}")
    print(f"Current 'Change Order?' value counts:\n{df['Change Order?'].value_counts()}")
else:
    print("Column 'Change Order?' not found in the DataFrame. No filtering applied.")

Filtered out rows where 'Change Order?' is 'Yes'. New shape: (3511, 45)
Updated DataFrame saved as: Cleaned_NBS_Orders_Targets.csv
Current 'Change Order?' value counts:
Change Order?
No    3499
Name: count, dtype: int64


In [19]:
df = pd.read_csv('Cleaned_NBS_Orders_Targets.csv')

In [20]:
df.head()

,Group,District,Region,Date,Month,Region Vlookup,District Vlookup,Branch Vlookup,Change Order Lookup,PROJECT,...,Unnamed: 35,Plan -2 Costs,Plan -2 CMII $,Plan -2 CMII%,MP?,CMII%.1,Change Order?,Product,DISTRICT,Branch Name
0,10381.0,Pacific,West,12/30/2025,12.0,ALL REGIONSAll OrdersNM12,PacificAll OrdersNM12,10381All OrdersNM12,All Orders,6865446.0,...,$-,"$234,061.20","$41,515.80",15.07,No,NaN,No,MonoSpace 500 4-6 landing(ENA),DPP,San Francisco
1,10130.0,Pacific,West,12/29/2025,12.0,ALL REGIONSAll OrdersNM12,PacificAll OrdersNM12,10130All OrdersNM12,All Orders,6867274.0,...,$-,"$214,947.44","$33,385.04",13.44,No,NaN,No,MonoSpace 500 4-6 landing(ENA),DPP,San Diego
2,10130.0,Pacific,West,12/29/2025,12.0,ALL REGIONSAll OrdersNM12,PacificAll OrdersNM12,10130All OrdersNM12,All Orders,6867274.0,...,$-,"$214,682.44","$33,656.04",13.55,No,NaN,No,MonoSpace 500 4-6 landing(ENA),DPP,San Diego
3,10130.0,Pacific,West,12/29/2025,12.0,ALL REGIONSAll OrdersNM12,PacificAll OrdersNM12,10130All OrdersNM12,All Orders,6867274.0,...,$-,"$215,370.05","$29,603.23",12.08,No,NaN,No,MonoSpace 500 4-6 landing(ENA),DPP,San Diego
4,10130.0,Pacific,West,12/29/2025,12.0,ALL REGIONSAll OrdersNM12,PacificAll OrdersNM12,10130All OrdersNM12,All Orders,6867274.0,...,$-,"$215,370.05","$29,603.23",12.08,No,NaN,No,MonoSpace 500 4-6 landing(ENA),DPP,San Diego


In [ ]:
print(df.shape[0])

3511


In [21]:
import time
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Filtrar filas de resumen y valores nulos
df_model = df.dropna(subset=['PROJECT NAME']).copy()

# 2. Preprocesamiento
features = ['District', 'Region', 'EQUIP TYPE', 'PROD CODE', 'Month', 'Plan -2 Costs']
X = pd.get_dummies(df_model[features], drop_first=True)
y = df_model['MV']

# 3. Modelo
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
rf = RandomForestRegressor(n_estimators=100, random_state=42)

start_time = time.time()
rf.fit(X_train, y_train)
y_pred = rf.predict(X_test)
end_time = time.time()
rf_execution_time = end_time - start_time

# 4. Evaluación
rf_r2 = r2_score(y_test, y_pred)
rf_mae = mean_absolute_error(y_test, y_pred)
rf_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
rf_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"R2 Score: {rf_r2:.4f}")
print(f"MAE: {rf_mae:.2f}")
print(f"RMSE: {rf_rmse:.2f}")
print(f"MAPE: {rf_mape:.2f}%")
print(f"Execution Time: {rf_execution_time:.2f} seconds")

# Store metrics
all_model_metrics.append({
    'Model': 'Random Forest Regressor',
    'R2 Score': rf_r2,
    'MAE': rf_mae,
    'RMSE': rf_rmse,
    'MAPE (%)': rf_mape,
    'Execution Time (s)': rf_execution_time
})


R2 Score: 0.8148
MAE: 33456.85
RMSE: 78066.06
MAPE: 11.42%
Execution Time: 22.47 seconds


Un R² de 0.8148 significa que el modelo explica aproximadamente el 81.5% de la variabilidad del precio de venta (MV)

Para el MAE = $33,456.85 tenemos un error v.s el promedio de:
139,928/ 33,456 ≈23.9%
Este error promedio absoluto es razonable para un problema de pricing con valores que superan los cien mil dólares.

RMSE = $78,066
Muestra que ya tenemos menos errores extremos, el modelo maneja mejor los casos dificiales y la prediccion en general mejoro

El MAPE de 11.42% indica que las predicciones del modelo se desvian este 11% del valor real. el cual resulta bastante solido

Esta mejora significa en el MAPE se dio una vez que indentificamos que muchos valores de MV en el dataset, o eran negativos o cercanos a cero o cercano a los millones. Estos datos no estan relacionados con los precios de venta sino con adiciones especiales no planeadas entre otros. Al eliminarlos del dataset base, los indicadores mejoraron significativamente


#**Modelo Alternativo 2 - Gradient Boosting  Regressor**

In [22]:
# Re-executing Gradient Boosting Regressor to collect metrics
import time
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# 1. Preparación de variables (X e y ya definidos en el paso anterior)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 2. Configuración del modelo (Equivalente a XGBoost)
gb_model = GradientBoostingRegressor(n_estimators=100,
                                     learning_rate=0.1,
                                     max_depth=5,
                                     random_state=42)

start_time = time.time()
# 3. Entrenamiento
gb_model.fit(X_train, y_train)

# 4. Predicción y Evaluación
y_pred = gb_model.predict(X_test)
end_time = time.time()
gb_execution_time = end_time - start_time

# Calculate metrics
gb_r2 = r2_score(y_test, y_pred)
gb_mae = mean_absolute_error(y_test, y_pred)
gb_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
gb_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"R2 Score: {gb_r2:.4f}")
print(f"MAE: {gb_mae:.2f}")
print(f"RMSE: {gb_rmse:.2f}")
print(f"MAPE: {gb_mape:.2f}%")
print(f"Execution Time: {gb_execution_time:.2f} seconds")

# 5. Ver importancia de variables
importances = pd.Series(gb_model.feature_importances_, index=X.columns).sort_values(ascending=False)
print("\nVariables más importantes:")
print(importances.head(5))

# Store metrics
all_model_metrics.append({
    'Model': 'Gradient Boosting Regressor',
    'R2 Score': gb_r2,
    'MAE': gb_mae,
    'RMSE': gb_rmse,
    'MAPE (%)': gb_mape,
    'Execution Time (s)': gb_execution_time
})


R2 Score: 0.8297
MAE: 37126.98
RMSE: 74848.61
MAPE: 14.68%
Execution Time: 13.74 seconds

Variables más importantes:
PROD CODE_ HR                0.375002
PROD CODE_ EM                0.115709
PROD CODE_ EX                0.096219
PROD CODE_ NH                0.072431
Plan -2 Costs_ $7,566.11     0.058158
dtype: float64


#**Modelo Alternativo 3 - Support Vector Regressor (SVR)**

In [23]:
# Re-executing Support Vector Regressor (SVR) to collect metrics
import time
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# 1. Preparar datos (usando las mismas variables X e y de antes)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 2. Crear un Pipeline (Escalamiento + Modelo)
svr_pipeline = make_pipeline(StandardScaler(), SVR(kernel='rbf', C=10000, epsilon=0.1))

start_time = time.time()
# 3. Entrenamiento
svr_pipeline.fit(X_train, y_train)

# 4. Predicción y Evaluación
y_pred = svr_pipeline.predict(X_test)
end_time = time.time()
svr_execution_time = end_time - start_time

# Calculate metrics
svr_r2 = r2_score(y_test, y_pred)
svr_mae = mean_absolute_error(y_test, y_pred)
svr_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
svr_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"SVR R2 Score: {svr_r2:.4f}")
print(f"SVR MAE: {svr_mae:.2f}")
print(f"SVR RMSE: {svr_rmse:.2f}")
print(f"SVR MAPE: {svr_mape:.2f}%")
print(f"Execution Time: {svr_execution_time:.2f} seconds")

# Store metrics
all_model_metrics.append({
    'Model': 'Support Vector Regressor (SVR)',
    'R2 Score': svr_r2,
    'MAE': svr_mae,
    'RMSE': svr_rmse,
    'MAPE (%)': svr_mape,
    'Execution Time (s)': svr_execution_time
})


SVR R2 Score: -0.1412
SVR MAE: 88213.74
SVR RMSE: 193760.56
SVR MAPE: 27.38%
Execution Time: 25.08 seconds


R² = -0.1412 es una señal clara de que el modelo no está capturando adecuadamente la relación entre las variables y el precio de venta.

#**Modelo alternativo 4 - MLP Regressor**

---



In [24]:
# Re-executing MLP Regressor to collect metrics
import time
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
import numpy as np

# 1. Preparación
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# 2. Escalador para y (target)
y_scaler = StandardScaler()

y_train_scaled = y_scaler.fit_transform(
    y_train.values.reshape(-1, 1)
).ravel()

# IMPORTANTE: solo transformamos test (no fit)
y_test_scaled = y_scaler.transform(
    y_test.values.reshape(-1, 1)
).ravel()

# 3. Modelo con escalado de X
mlp_model = Pipeline([
    ('scaler', StandardScaler()),
    ('mlp', MLPRegressor(
        hidden_layer_sizes=(100, 50),
        max_iter=2000,
        activation='relu',
        solver='adam',
        random_state=42
    ))
])

start_time = time.time()
# 4. Entrenamiento (usando y escalado)
mlp_model.fit(X_train, y_train_scaled)

# 5. Predicción (en escala normal)
y_pred_scaled = mlp_model.predict(X_test)

# volver a escala original
y_pred = y_scaler.inverse_transform(
    y_pred_scaled.reshape(-1, 1)
).ravel()
end_time = time.time()
mlp_execution_time = end_time - start_time

# 6. Evaluación
mlp_r2 = r2_score(y_test, y_pred)
mlp_mae = mean_absolute_error(y_test, y_pred)
mlp_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
mlp_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"Red Neuronal R2: {mlp_r2:.4f}")
print(f"Red Neuronal MAE: {mlp_mae:.2f}")
print(f"Red Neuronal RMSE: {mlp_rmse:.2f}")
print(f"Red Neuronal MAPE: {mlp_mape:.2f}%")
print(f"Execution Time: {mlp_execution_time:.2f} seconds")

# Store metrics
all_model_metrics.append({
    'Model': 'MLP Regressor',
    'R2 Score': mlp_r2,
    'MAE': mlp_mae,
    'RMSE': mlp_rmse,
    'MAPE (%)': mlp_mape,
    'Execution Time (s)': mlp_execution_time
})


Red Neuronal R2: -1.6202
Red Neuronal MAE: 243366.11
Red Neuronal RMSE: 293595.61
Red Neuronal MAPE: 140.65%
Execution Time: 17.42 seconds


#**Modelo alternativo 5 - XGBoost Regressor**

---


In [25]:
# Re-executing XGBoost Regressor to collect metrics
import time
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
import numpy as np
import pandas as pd

# 1. Preparación de variables (X e y ya definidos)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# 2. Configuración del modelo XGBoost
xgb_model = XGBRegressor(
    n_estimators=200,
    learning_rate=0.1,
    max_depth=5,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42
)

start_time = time.time()
# 3. Entrenamiento
xgb_model.fit(X_train, y_train)

# 4. Predicción
y_pred = xgb_model.predict(X_test)
end_time = time.time()
xgb_execution_time = end_time - start_time

# 5. Métricas
xgb_r2 = r2_score(y_test, y_pred)
xgb_mae = mean_absolute_error(y_test, y_pred)
xgb_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
xgb_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"R2 Score: {xgb_r2:.4f}")
print(f"MAE: {xgb_mae:.2f}")
print(f"RMSE: {xgb_rmse:.2f}")
print(f"MAPE: {xgb_mape:.2f}%")
print(f"Execution Time: {xgb_execution_time:.2f} seconds")

# 6. Importancia de variables
importances = pd.Series(
    xgb_model.feature_importances_,
    index=X.columns
).sort_values(ascending=False)

print("\nVariables más importantes:")
print(importances.head(5))

# Store metrics
all_model_metrics.append({
    'Model': 'XGBoost Regressor',
    'R2 Score': xgb_r2,
    'MAE': xgb_mae,
    'RMSE': xgb_rmse,
    'MAPE (%)': xgb_mape,
    'Execution Time (s)': xgb_execution_time
})


R2 Score: 0.8427
MAE: 35011.80
RMSE: 71929.95
MAPE: 13.63%
Execution Time: 8.36 seconds

Variables más importantes:
PROD CODE_ HR                0.147426
Plan -2 Costs_ $7,566.11     0.076120
PROD CODE_ EX                0.060080
PROD CODE_ NE                0.059275
PROD CODE_ EM                0.056791
dtype: float32


#**Modelo alternativo 6 - KNN**

---


In [26]:
# Re-executing K-Nearest Neighbors (KNN) to collect metrics
import time
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
import numpy as np
import pandas as pd

df_model = df.dropna(subset=['PROJECT NAME']).copy()

# Clean 'Plan -2 Costs' to ensure it's numeric before using it in X
if 'Plan -2 Costs' in df_model.columns:
    df_model['Plan -2 Costs'] = df_model['Plan -2 Costs'].astype(str).str.replace('$', '', regex=False).str.replace(',', '', regex=False)
    df_model['Plan -2 Costs'] = df_model['Plan -2 Costs'].apply(lambda x: '-' + x[1:-1] if isinstance(x, str) and x.startswith('(') and x.endswith(')') else x)
    df_model['Plan -2 Costs'] = pd.to_numeric(df_model['Plan -2 Costs'], errors='coerce')
    df_model['Plan -2 Costs'] = df_model['Plan -2 Costs'].fillna(df_model['Plan -2 Costs'].median()) # Fixed: Removed inplace=True

features_to_encode = ['District', 'Region', 'EQUIP TYPE', 'PROD CODE'] # Categorical features
numerical_features = ['Month', 'Plan -2 Costs'] # Numerical features

X_categorical = pd.get_dummies(df_model[features_to_encode], drop_first=True)
X_numerical = df_model[numerical_features]
X = pd.concat([X_categorical, X_numerical], axis=1)
y = df_model['MV']

# 1. Preparar datos
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# 2. Pipeline (Escalamiento + KNN)
knn_pipeline = make_pipeline(
    StandardScaler(),
    KNeighborsRegressor(
        n_neighbors=5,
        weights='distance'
    )
)

start_time = time.time()
# 3. Entrenamiento
knn_pipeline.fit(X_train, y_train)

# 4. Predicción
y_pred = knn_pipeline.predict(X_test)
end_time = time.time()
knn_execution_time = end_time - start_time

# 5. Métricas básicas
knn_r2 = r2_score(y_test, y_pred)
knn_mae = mean_absolute_error(y_test, y_pred)
knn_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# The mean_absolute_percentage_error function is now defined globally in RZhopTnONk1z
knn_mape = mean_absolute_percentage_error(y_test, y_pred)

print(f"KNN R2 Score: {knn_r2:.4f}")
print(f"KNN MAE: {knn_mae:.2f}")
print(f"KNN RMSE: {knn_rmse:.2f}")
print(f"KNN MAPE: {knn_mape:.2f}%")
print(f"Execution Time: {knn_execution_time:.2f} seconds")

# Store metrics
all_model_metrics.append({
    'Model': 'K-Nearest Neighbors (KNN)',
    'R2 Score': knn_r2,
    'MAE': knn_mae,
    'RMSE': knn_rmse,
    'MAPE (%)': knn_mape,
    'Execution Time (s)': knn_execution_time
})


KNN R2 Score: 0.9231
KNN MAE: 14723.33
KNN RMSE: 50281.81
KNN MAPE: 5.13%
Execution Time: 0.03 seconds


El modelo K-Nearest Neighbors obtuvo un desempeño sobresaliente, alcanzando un R² de 0.9231 y un MAPE de 5.13%, lo que indica una alta capacidad para capturar similitudes locales en los datos. Estos resultados sugieren que el comportamiento del precio de venta presenta patrones altamente estructurados donde observaciones similares tienden a tener valores de venta cercanos. Sin embargo, debido a la naturaleza del algoritmo basado en distancias, estos resultados deben ser validados cuidadosamente para descartar posibles efectos de similitud excesiva entre observaciones o fugas de información.

#**Comparacion de rendimiento**

---


In [27]:
import pandas as pd

# Create a DataFrame from the collected metrics
metrics_df = pd.DataFrame(all_model_metrics)

# Display the DataFrame
print(metrics_df.to_markdown(index=False))


| Model                          |   R2 Score |      MAE |     RMSE |   MAPE (%) |   Execution Time (s) |
|:-------------------------------|-----------:|---------:|---------:|-----------:|---------------------:|
| Random Forest Regressor        |   0.814753 |  33456.8 |  78066.1 |   11.4245  |           22.4659    |
| Gradient Boosting Regressor    |   0.829708 |  37127   |  74848.6 |   14.6821  |           13.7436    |
| Support Vector Regressor (SVR) |  -0.14119  |  88213.7 | 193761   |   27.3756  |           25.0761    |
| MLP Regressor                  |  -1.62015  | 243366   | 293596   |  140.652   |           17.4197    |
| XGBoost Regressor              |   0.84273  |  35011.8 |  71930   |   13.6258  |            8.35833   |
| K-Nearest Neighbors (KNN)      |   0.923149 |  14723.3 |  50281.8 |    5.12589 |            0.0256481 |


R2:El modelo KNN presenta un R² de 0.9231, lo que indica una capacidad explicativa muy alta, logrando capturar aproximadamente el 92% de la variabilidad del precio de venta. En comparación, XGBoost obtiene un R² de 0.8427, que aunque sigue siendo un desempeño sólido y muy competitivo, queda por debajo del KNN. Esta diferencia sugiere que KNN está aprovechando muy bien la estructura local de los datos, mientras que XGBoost captura patrones globales más generales pero con menor ajuste fino a las observaciones individuales.

MAE:En términos de MAE, KNN obtiene 14,723, significativamente menor que los 35,011 de XGBoost. Esto indica que, en promedio, las predicciones de KNN se desvían mucho menos del valor real en unidades monetarias, lo cual es una ventaja clara en precisión directa. Por otro lado, XGBoost presenta un error absoluto más alto, lo que sugiere que, aunque es un modelo robusto, tiende a cometer errores más grandes en promedio comparado con KNN.

RSME:El RMSE de KNN es de 50,281, considerablemente inferior al de XGBoost, que es 71,929. Esto indica que KNN no solo tiene errores promedio más bajos, sino también menos errores extremos o outliers significativos en sus predicciones. En contraste, XGBoost muestra una mayor penalización por errores grandes, lo que sugiere que tiene más dificultad en ciertos casos específicos del dataset donde las predicciones se desvían considerablemente.

MAPE: El MAPE de KNN es de 5.13%, lo cual es extremadamente bajo y representa un desempeño excelente en términos relativos, indicando que las predicciones se desvían en promedio solo un 5% del valor real. XGBoost, con un MAPE de 13.63%, sigue siendo un modelo aceptable y bueno para problemas de negocio, pero claramente menos preciso en términos porcentuales. Esta diferencia refuerza la superioridad de KNN en precisión relativa, especialmente en un contexto de forecasting de precios.


#**Ajuste de parametros de los dos mejores modelos**    

---


Se ajustan los parametros para el modelo XGBoost y se obtiene un R2 de 0.95

In [29]:
import time
import numpy as np
import pandas as pd
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split, RandomizedSearchCV, KFold
from sklearn.metrics import (
    r2_score, mean_absolute_error,
    mean_squared_error, mean_absolute_percentage_error
)
from scipy.stats import randint, uniform


X_train_full, X_test, y_train_full, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


X_train, X_val, y_train, y_val = train_test_split(
    X_train_full, y_train_full, test_size=0.2, random_state=42
)


param_dist = {
    "n_estimators":      randint(300, 1200),
    "learning_rate":     uniform(0.01, 0.19),
    "max_depth":         randint(3, 9),
    "min_child_weight":  randint(1, 8),
    "gamma":             uniform(0, 0.5),
    "subsample":         uniform(0.6, 0.4),
    "colsample_bytree":  uniform(0.6, 0.4),
    "reg_alpha":         uniform(0, 1.0),
    "reg_lambda":        uniform(0.5, 2.0),
}

base_model = XGBRegressor(
    objective="reg:squarederror",
    tree_method="hist",
    random_state=42,
    n_jobs=-1,
)

cv = KFold(n_splits=5, shuffle=True, random_state=42)

search = RandomizedSearchCV(
    estimator=base_model,
    param_distributions=param_dist,
    n_iter=60,
    scoring="neg_root_mean_squared_error",
    cv=cv,
    verbose=1,
    random_state=42,
    n_jobs=-1,
)


start_time = time.time()
search.fit(X_train_full, y_train_full)
best_params = search.best_params_
print("Mejores hiperparámetros:")
for k, v in best_params.items():
    print(f"  {k}: {v}")

# early stopping
xgb_model = XGBRegressor(
    **best_params,
    objective="reg:squarederror",
    tree_method="hist",
    random_state=42,
    n_jobs=-1,
    early_stopping_rounds=50,
    eval_metric="rmse",
)
xgb_model.fit(
    X_train, y_train,
    eval_set=[(X_val, y_val)],
    verbose=False,
)
print(f"\nMejor iteración (early stopping): {xgb_model.best_iteration}")


y_pred = xgb_model.predict(X_test)
end_time = time.time()
xgb_execution_time = end_time - start_time

xgb_r2   = r2_score(y_test, y_pred)
xgb_mae  = mean_absolute_error(y_test, y_pred)
xgb_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
xgb_mape = mean_absolute_percentage_error(y_test, y_pred) * 100

print(f"\nR2 Score: {xgb_r2:.4f}")
print(f"MAE: {xgb_mae:.2f}")
print(f"RMSE: {xgb_rmse:.2f}")
print(f"MAPE: {xgb_mape:.2f}%")
print(f"Execution Time: {xgb_execution_time:.2f} seconds")


importances = pd.Series(
    xgb_model.feature_importances_,
    index=X.columns
).sort_values(ascending=False)
print("\nVariables más importantes:")
print(importances.head(5))

# Métricas
all_model_metrics.append({
    "Model": "XGBoost Regressor (tuned)",
    "R2 Score": xgb_r2,
    "MAE": xgb_mae,
    "RMSE": xgb_rmse,
    "MAPE (%)": xgb_mape,
    "Execution Time (s)": xgb_execution_time,
})

Fitting 5 folds for each of 60 candidates, totalling 300 fits
Mejores hiperparámetros:
  colsample_bytree: 0.754158648386226
  gamma: 0.34080683448639487
  learning_rate: 0.0747186681982056
  max_depth: 5
  min_child_weight: 1
  n_estimators: 495
  reg_alpha: 0.6928903586919394
  reg_lambda: 1.1966732089063947
  subsample: 0.9746592607422829

Mejor iteración (early stopping): 238

R2 Score: 0.9454
MAE: 12982.35
RMSE: 42398.19
MAPE: 4.84%
Execution Time: 123.65 seconds

Variables más importantes:
PROD CODE_ HR     0.600161
Plan -2 Costs     0.142314
PROD CODE_ EM     0.089843
PROD CODE_ EX     0.037818
PROD CODE_ NH     0.026143
dtype: float32


Se ajustan hiperparametros de KNN y se obtiene un R2 de 0.92

In [30]:
import time
import numpy as np
import pandas as pd
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV, KFold
from sklearn.metrics import (
    r2_score, mean_absolute_error,
    mean_squared_error, mean_absolute_percentage_error
)


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


knn_pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("knn", KNeighborsRegressor()),
])


param_grid = {
    "knn__n_neighbors": list(range(3, 31, 2)),
    "knn__weights":     ["uniform", "distance"],
    "knn__p":           [1, 2],
    "knn__metric":      ["minkowski"],
}

cv = KFold(n_splits=5, shuffle=True, random_state=42)

search = GridSearchCV(
    estimator=knn_pipeline,
    param_grid=param_grid,
    scoring="neg_root_mean_squared_error",
    cv=cv,
    verbose=1,
    n_jobs=-1,
)


start_time = time.time()
search.fit(X_train, y_train)

best_params = search.best_params_
print("Mejores hiperparámetros:")
for k, v in best_params.items():
    print(f"  {k}: {v}")

knn_model = search.best_estimator_


y_pred = knn_model.predict(X_test)
end_time = time.time()
knn_execution_time = end_time - start_time

knn_r2   = r2_score(y_test, y_pred)
knn_mae  = mean_absolute_error(y_test, y_pred)
knn_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
knn_mape = mean_absolute_percentage_error(y_test, y_pred) * 100

print(f"\nKNN R2 Score: {knn_r2:.4f}")
print(f"KNN MAE: {knn_mae:.2f}")
print(f"KNN RMSE: {knn_rmse:.2f}")
print(f"KNN MAPE: {knn_mape:.2f}%")
print(f"Execution Time: {knn_execution_time:.2f} seconds")

# Métricas
all_model_metrics.append({
    "Model": "K-Nearest Neighbors (KNN, tuned)",
    "R2 Score": knn_r2,
    "MAE": knn_mae,
    "RMSE": knn_rmse,
    "MAPE (%)": knn_mape,
    "Execution Time (s)": knn_execution_time,
})

Fitting 5 folds for each of 56 candidates, totalling 280 fits
Mejores hiperparámetros:
  knn__metric: minkowski
  knn__n_neighbors: 3
  knn__p: 2
  knn__weights: distance

KNN R2 Score: 0.9215
KNN MAE: 14469.28
KNN RMSE: 50825.28
KNN MAPE: 5.11%
Execution Time: 11.82 seconds


#**Justificacion modelo final**
---

# **Los resultados obtenidos muestran que el modelo K-Nearest Neighbors presenta el mejor desempeño en términos de precisión, alcanzando un R² de 0.923 y el menor error absoluto y relativo entre los modelos evaluados. Posiblemente por la estructura del dataset con fuerte agrupación por similud y precios dependientes de casos cercanos o también por baja complejidad. EL problema con esto es que un modelo KNN pudiera no generalizar bien fuera de los datos conocidos y pudiera tener peor desempeño si la distribución cambiara**

# **Por el contrario, modelos basados en ensambles como XGBoost y Random Forest ofrecen un mejor balance entre precisión, robustez y capacidad de generalización. En particular, XGBoost destaca como el modelo más adecuado para aplicaciones prácticas debido a su alto desempeño y estabilidad. Por otro lado, modelos como SVR y redes neuronales no lograron capturar adecuadamente la estructura del problema, presentando incluso desempeño inferior a un modelo base.**

# **Al respecto del tiempo de procesamiento: KNN es rápido en entrenamiento porque no entrena un modelo explícito, sino que almacena los datos y realiza el cálculo de distancias únicamente en el momento de la predicción.Ademas nuestro dataset resulta ser de un tamaño relativamente pequeño.En el caso de los demás modelos, se toman tiempo para construccion de los arboles (Random Forest), entrenamiento sequencial  (XGB) o backpropagation (MLP)**